sql/compressed_uint64_fifo.h

```diff
@@ -0,0 +1,227 @@
+    }
+
+    void advance() noexcept {
+      if (stream_itr == container.data.end()) {
+        if (used_last_run_of_ones_len == container.current_run_of_ones_len) {
+          assert(current_number == container.last_number);
+          current_number = 0;
+          return;
+        }
+        ++used_last_run_of_ones_len;
+        ++current_number;
+        return;
+      }
+      if (remaining_run_of_ones_len > 0) {
+        ++current_number;
+        --remaining_run_of_ones_len;
+        return;
+      }
+      const auto [number, is_run_of_ones_length] = read();
+      if (is_run_of_ones_length) {
+        ++current_number;
+        remaining_run_of_ones_len = number - 1;
+        return;
+      }
+      assert(remaining_run_of_ones_len == 0);
+      current_number += number;
+    }
+
+    friend class compressed_uint64_fifo;
+
+    // Return iterator that points at the end
+    const_iterator(const compressed_uint64_fifo &container_, bool) noexcept
+        : container{container_},
+          stream_itr{container.data.cend()},
+          used_last_run_of_ones_len{container.current_run_of_ones_len} {}
+
+    explicit const_iterator(const compressed_uint64_fifo &container_) noexcept
+        : container{container_},
+          stream_itr{container.data.cbegin()},
+          used_last_run_of_ones_len{0} {
+      advance();
+    }
+
+   public:
+    [[nodiscard]] std::uint64_t operator*() const noexcept {
+      return current_number;
+    }
+
+    [[nodiscard]] const_iterator &operator++() noexcept {
+      advance();
+      return *this;
+    }
+
+    [[nodiscard]] bool operator==(const const_iterator &other) const noexcept {
+      assert(&container == &other.container);
+      return current_number == other.current_number;
+    }
+
+    [[nodiscard]] bool operator!=(const const_iterator &other) const noexcept {
+      assert(&container == &other.container);
+      return current_number != other.current_number;
+    }
+  };
+
+  [[nodiscard]] bool empty() const noexcept {
+    return data.empty() && current_run_of_ones_len == 0;
+  }
+
+  [[nodiscard]] std::uint64_t back() const noexcept {
+    assert(!empty());
+    return last_number;
+  }
+
+  void clear() {
+    data.clear();
+    last_number = 0;
+    current_run_of_ones_len = 0;
+  }
+
+  void push_back(std::uint64_t value) {
+    // MyISAM supporting auto_increment on key parts means that values may go
+    // down as well, breaking the monotonicity. Support this by allowing large
+    // deltas that wrap around. The compression will be poor but MyISAM is not
+    // used much (or at all).
+    // Rethink the approach if MyRocks starts supporting composite primary keys
+    // with auto_increment key suffixes. Otherwise replace this assert with a
+    // ">" one once MyISAM is killed.
+    assert(value != last_number);
+
+    auto delta = value - last_number;
+    if (delta == 1) {
+      ++current_run_of_ones_len;
+    } else {
+      if (current_run_of_ones_len > 0) {
```

Comment:
I worried that unconditional reserve(128) will be wasteful in too many cases, where we only have a single value inserted and when the deltas are only equal to 1, and we don't touch the vector at all. Maybe we can put `reserve(128)` into this else branch. The first time it runs, it will do that, and will be a no-op afterwards.